// @gluestack-ui/pressable/lib/commonjs/Pressable.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _focus = require("@react-native-aria/focus");
var _react = _interopRequireWildcard(require("react"));
var _interactions = require("@react-native-aria/interactions");
var _utils = require("@gluestack-ui/utils");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function Pressable(StyledPressable) {
  return /*#__PURE__*/(0, _react.forwardRef)(({
    children,
    ...props
  }, ref) => {
    const {
      focusProps: focusRingProps,
      isFocusVisible
    } = (0, _focus.useFocusRing)();
    const {
      pressProps,
      isPressed
    } = (0, _interactions.usePress)({
      isDisabled: props.disabled ?? undefined
    });
    const {
      isFocused,
      focusProps
    } = (0, _focus.useFocus)();
    const {
      isHovered,
      hoverProps
    } = (0, _interactions.useHover)();
    return /*#__PURE__*/_react.default.createElement(StyledPressable, _extends({
      ref: ref,
      states: {
        hover: isHovered,
        focus: isFocused,
        active: isPressed,
        disabled: props.disabled,
        focusVisible: isFocusVisible
      },
      dataSet: {
        hover: isHovered ? 'true' : 'false',
        focus: isFocused ? 'true' : 'false',
        active: isPressed ? 'true' : 'false',
        disabled: props.disabled ? 'true' : 'false',
        focusVisible: isFocusVisible ? 'true' : 'false'
      }
    }, props, {
      onPressIn: (0, _utils.composeEventHandlers)(props === null || props === void 0 ? void 0 : props.onPressIn, pressProps.onPressIn),
      onPressOut: (0, _utils.composeEventHandlers)(props === null || props === void 0 ? void 0 : props.onPressOut, pressProps.onPressOut)
      // @ts-ignore - web only
      ,
      onHoverIn: (0, _utils.composeEventHandlers)(props === null || props === void 0 ? void 0 : props.onHoverIn, hoverProps.onHoverIn)
      // @ts-ignore - web only
      ,
      onHoverOut: (0, _utils.composeEventHandlers)(props === null || props === void 0 ? void 0 : props.onHoverOut, hoverProps.onHoverOut)
      // @ts-ignore - web only
      ,
      onFocus: (0, _utils.composeEventHandlers)((0, _utils.composeEventHandlers)(props === null || props === void 0 ? void 0 : props.onFocus, focusProps.onFocus), focusRingProps.onFocus)
      // @ts-ignore - web only
      ,
      onBlur: (0, _utils.composeEventHandlers)((0, _utils.composeEventHandlers)(props === null || props === void 0 ? void 0 : props.onBlur, focusProps.onBlur), focusRingProps.onBlur)
    }), typeof children === 'function' ? children({
      hovered: isHovered,
      focused: isFocused,
      pressed: isPressed,
      disabled: props.disabled ?? undefined,
      focusVisible: isFocusVisible
    }) : children);
  });
}
var _default = exports.default = Pressable;
