// @gluestack-ui/tabs/lib/module/TabPanel.js
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
import React, { memo } from 'react';
import { forwardRef } from 'react';
import { Platform } from 'react-native';
import { useTab } from './TabProvider';
export const TabPanel = StyledTabPanel => /*#__PURE__*/memo( /*#__PURE__*/forwardRef(({
  value,
  ...props
}, ref) => {
  const {
    currentActiveTab
  } = useTab('TabContext');
  const isActive = value === currentActiveTab;
  if (isActive) return /*#__PURE__*/React.createElement(StyledTabPanel, _extends({
    tabIndex: value === currentActiveTab ? 0 : -1,
    role: Platform.OS === 'web' ? 'tabpanel' : undefined
  }, props, {
    style: {
      display: isActive ? 'flex' : 'none'
    },
    ref: ref
  }));
  return /*#__PURE__*/React.createElement(React.Fragment, null);
}));
